src/main/java/net/neoforged/neoforge/common/crafting/ICustomIngredient.java

```diff
@@ -0,0 +1,76 @@
+ */
+
+package net.neoforged.neoforge.common.crafting;
+
+import java.util.stream.Stream;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.neoforged.neoforge.registries.NeoForgeRegistries;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Interface that modders can implement to create new behaviors for {@link Ingredient}s.
+ *
+ * <p>This is not directly implemented on vanilla {@link Ingredient}s, but conversions are possible:
+ * <ul>
+ * <li>{@link #toVanilla()} converts a custom ingredient to a vanilla {@link Ingredient}.</li>
+ * <li>{@link Ingredient#getCustomIngredient()} retrieves the custom ingredient inside a vanilla {@link Ingredient}.</li>
+ * </ul>
+ *
+ * <p>Implementations of this interface <b>must implement {@link Object#equals} and {@link Object#hashCode}</b>
+ * to ensure that the ingredient also supports them.
+ */
+public interface ICustomIngredient {
+    /**
+     * Checks if a stack matches this ingredient.
+     * The stack <b>must not</b> be modified in any way.
+     *
+     * @param stack the stack to test
+     * @return {@code true} if the stack matches this ingredient, {@code false} otherwise
+     */
+    boolean test(ItemStack stack);
+
+    /**
+     * {@return the list of stacks that this ingredient accepts}
+     *
+     * <p>The following guidelines should be followed for good compatibility:
+     * <ul>
+     * <li>These stacks are generally used for display purposes, and need not be exhaustive or perfectly accurate.</li>
+     * <li>An exception is ingredients that {@linkplain #isSimple() are simple},
+     * for which it is important that the returned stacks correspond exactly to all the accepted {@link Item}s.</li>
+     * <li>At least one stack must be returned for the ingredient not to be considered {@linkplain Ingredient#hasNoItems() accidentally empty}.</li>
+     * <li>The ingredient should try to return at least one stack with each accepted {@link Item}.
+     * This allows mods that inspect the ingredient to figure out which stacks it might accept.</li>
+     * </ul>
+     *
+     * <p>Note: no caching needs to be done by the implementation, this is already handled by the ingredient itself.
+     */
+    Stream<ItemStack> getItems();
+
+    /**
+     * Returns whether this ingredient always requires {@linkplain #test direct stack testing}.
+     *
+     * @return {@code true} if this ingredient ignores NBT data when matching stacks, {@code false} otherwise
+     * @see Ingredient#isSimple()
+     */
+    boolean isSimple();
+
+    /**
+     * {@return the type of this ingredient}
+     *
+     * <p>The type must be registered to {@link NeoForgeRegistries#INGREDIENT_TYPES}.
+     */
+    IngredientType<?> getType();
+
+    /**
+     * {@return a new {@link Ingredient} behaving as defined by this custom ingredient}
+     */
+    @ApiStatus.NonExtendable
+    default Ingredient toVanilla() {
```

Comment:
Is a `toSized(int)` method worth having here?

Reply:
I don't think so. Generally the first thing you'll do with an `ICustomIngredient` is turn it into an `Ingredient`. That's what all our static `XxxIngredient.of(...)` methods do.

Reply:
Alternatively adding a `SizedIngredient.of(ICustomIngredient,int)` method would be useful. Just so you don't always have to call `toVanilla` to construct sized

Reply:
I don't think you're likely to directly obtain an instance of an `ICustomIngredient`? You'll get an `Ingredient` most of the time even if it is backed by an `ICustomIngredient`.

Reply:
I'm just talking about datagen. If I want to create say, a sized difference ingredient, I'd have to do `SizedIngredient.of(new Difference ingredient(whateverArguments).toVanilla(), 5)`. Creating sized custom ingredients seems like it will be common enough that it be nice to leave off the `toVanilla` here

Reply:
You're not supposed to call the custom constructors directly, and I'm pretty sure the API specifies that as well? At the very least, all the custom Ingredients have static `of` methods that call `.toVanilla()` automatically

Reply:
If thats the case, then I don't see why we need `toVanilla()` at all; if the API design is just make and use the static methods, there is no reason the static methods can't just call the static method on `Ingredient`. All it does is add an overridable method that you are not supposed to call or override, which is pretty pointless API.